pkg/scheduler/plugins/proportion/resource_division/priority_resource_division.go

```diff
@@ -0,0 +1,81 @@
+package resource_division
+
+import (
+	"math"
+
+	commonconstants "github.com/NVIDIA/KAI-scheduler/pkg/common/constants"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/common_info"
+	rs "github.com/NVIDIA/KAI-scheduler/pkg/scheduler/plugins/proportion/resource_share"
+)
+
+// SetResourcesSharePriority distributes resources between queues according to the
+// three-phase algorithm described in docs/developer/designs/priority-based-fair-share.md.
+//
+//  1. Queues are grouped by priority (higher priority buckets processed first).
+//  2. Within a bucket we run:
+//     a) Phase-1: allocate Deserved quota (subject to remaining capacity).
+//     b) Phase-2: allocate over-quota resources proportionally to OverQuotaWeight.
+//     c) Phase-3: allocate any still-remaining capacity equally between the queues
+//     that still have unmet demand (treating all weights as 1).
+func SetResourcesSharePriority(totalResource rs.ResourceQuantities, queues map[common_info.QueueID]*rs.QueueAttributes) {
+	for _, resource := range rs.AllResources {
+		remaining := totalResource[resource]
+		if remaining <= 0 {
+			continue
+		}
+
+		// Group queues by priority (higher first).
+		queuesByPriority, priorities := getQueuesByPriority(queues)
+
+		for _, priority := range priorities {
+			if remaining <= 0 {
+				break
+			}
+			bucket := queuesByPriority[priority]
+
+			// ---------- Phase 1: Deserved quota ----------
+			for _, q := range bucket {
+				if remaining <= 0 {
+					break
+				}
+				rsShare := q.ResourceShare(resource)
```

Comment:
This is too deep! Let's keep the indentations low - this loop could be offloaded to a function

Reply:
:+1: I will refactor it into several functions.